[onert] Revise Event and EventWriter classes

hyunsik-yoon opened the pull request:
This revises `Event` and `EventWriter` classes.

Parent issue: #4901
Draft: #4903

This is the last PR for the draft.

Sorry about large PR but it's kinda hard to split this into several PRs.

Major changes are:
1. Refine `Event` class (`EventCollector.h`) to have two child classes: `SubgEvent` and `OpEvent`
2. Refine `DurationEvent` class (`DurationEventRecorder.h`) 
    - to have two child classes: `SubgDurationEvent` and `OpDurationEvent`
    - to have higher member vars. Previously, it had, e.g., `tid` (thread id), which is specific to Chrome Tracing. However `tid` is not meaningful to `MDEventWriter` or `SNPEEventWriter`
3. Make specific event writer decide labels (and tid)
   - So writers (`ChromeTracingEventWriter`, `MDEventWriter` and `SNPEEventWriter`) now have, its own, e.g., `LabelMaker`
4. Add session and subgraph index into label

I will also leave some explanations in code.

Signed-off-by: Hyun Sik Yoon <hyunsik.yoon.1024@gmail.com>

hyunsik-yoon commented:
Please also refer to the result log files below:
- [logs.tar.gz](https://github.com/Samsung/ONE/files/5700640/logs.tar.gz)

Some screen shots are as follows:

---------

```
$ TRACE_FILEPATH=../GenModelTest.OneOp_While_github_4783.log OP_SEQ_MAX_NODE=1  ./Product/out/unittest_standalone/nnfw_api_gtest --gtest_filter=*.OneOp_While_github_4783
```

![image](https://user-images.githubusercontent.com/63768207/102311311-feb47c80-3faf-11eb-940e-04aed2af4224.png)


----------

```
$ TRACE_FILEPATH=../ValidationTestTwoSessionsCreated.two_sessions_run_simple_While_model.log OP_SEQ_MAX_NODE=1  ./Product/out/unittest_standalone/nnfw_api_gtest --gtest_filter=ValidationTestTwoSessionsCreated.two_sessions_run_simple_While_model
```

![image](https://user-images.githubusercontent.com/63768207/102311033-78983600-3faf-11eb-934e-4fffbf81d2aa.png)

--------------

```
$ TRACE_FILEPATH=../mobilenet.log OP_SEQ_MAX_NODE=1  ./Product/out/bin/nnpackage_run --nnpackage ../../4/mobilenet_v1_0.25_128
```

![image](https://user-images.githubusercontent.com/63768207/102311077-8e0d6000-3faf-11eb-94bb-5db50b120b6a.png)

hyunsik-yoon commented on runtime/onert/core/include/util/TracingCtx.h:
```diff
@@ -63,6 +63,13 @@ class TracingCtx
+   *
+   * @note  This method is NOT thread-safe. Call this in thread-safe situation.
+   */
+  bool hasMultipleSessions() const { return _next_session_id > 1; }
```
if there are multiple session, `SNPEEventWriter` will use `"$1 session $2 subgraph $23 Add"` as op name.
However if there is only one session, `SNPEEventWriter` will use `"$2 subgraph $23 Add"` without `"$0 session"`. This is requested by Tizen people.
This method is to check if there are multiple session.

hyunsik-yoon commented on runtime/onert/core/src/exec/ExecutionObservers.cc:
```diff
@@ -146,61 +142,43 @@ TracingObserver::~TracingObserver()
-  _collector.onEvent(EventCollector::Event{EventCollector::Edge::END, "runtime", "Graph"});
-}
-
-std::string TracingObserver::opSequenceTag(const ir::OpSequence *op_seq,
```
Removed since this is meaningful when making the name of op for chrome tracing. Same feature was added into ChromeTracingEventWriter.

hyunsik-yoon commented on runtime/onert/core/src/util/ChromeTracingEventWriter.cc:
```diff
@@ -115,6 +122,45 @@ std::string object(const CounterEvent &evt)
+  else
+    return "$" + std::to_string(evt.op_index) + " " + evt.op_name;
+}
+
```
`LabelMaker` makes name of op specific for ChromeTracingEventWriter.
`TidMaker` makes `tid` row of for ChromeTracing screen.

hyunsik-yoon commented on runtime/onert/core/src/util/EventRecorder.h:
```diff
@@ -17,28 +17,75 @@
+// refer to https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/edit#
 struct Event
 {
-  std::string name;
```
`name` and `tid` are different from type of writer(e.g., `ChromeTracingEventWriter`). So now deciding `name` and `tid` is specific writer's role. Thus, they are removed from here.

hyunsik-yoon commented on runtime/onert/core/src/util/EventRecorder.h:
```diff
@@ -17,28 +17,75 @@
+
+  std::string accept(DurationEventVisitor &visitor) const override { return visitor.visit(*this); }
 };
 
```
With this PR, I tried NOT to touch any memory measuring code including `CounterEvent`. So it has almost same member variables as before.

hyunsik-yoon commented on runtime/onert/core/src/util/MDTableEventWriter.cc:
```diff
@@ -178,9 +180,22 @@ struct Graph : public MDContent
   }
 };
 
```
making op name for `MDTableEventWriter`

wateret commented on runtime/onert/core/src/util/EventRecorder.h:
```diff
@@ -17,28 +17,75 @@
+  std::string accept(DurationEventVisitor &visitor) const override { return visitor.visit(*this); }
+};
+
+struct OpDurationEvent : public DurationEvent
```
Is this supposed to mean `OpSequence` duration or `Operation`?

hyunsik-yoon replied:
Yes. maybe it would better to rename it to `OpSeqDurationEvent`. I will fix.

wateret commented on runtime/onert/core/src/util/EventRecorder.h:
```diff
@@ -17,28 +17,75 @@
+  {
+    throw std::runtime_error("Please implement");
+  }
+};
```
(Opinion-based) IMHO introducing a visitor for DurationEvent is an overkill since:

- There are only two events
- What visitors do are relatively simple
- Changing event-recording module does not affect other modules

I'm not saying it is unacceptable, though.

hyunsik-yoon replied:
I will change visitors to two class architecture.

wateret commented on runtime/onert/core/src/util/ChromeTracingEventWriter.cc:
```diff
@@ -115,6 +122,45 @@ std::string object(const CounterEvent &evt)
 
+std::string getSessionLabel(const DurationEvent &evt)
+{
+  return "$" + std::to_string(evt.session_index) + " sess";
```
We may change the symbols later 😅

chunseoklee reviewed (dismissed):
LGTM for me

hyunsik-yoon commented:
Fixed. PTAL.

- I removed visitor and wrote separate functions to handle `SubgEvent` and `OpSeqEvent` (and also `SubgDurationEvent` and `OpSeqDurationEvent`)
- I also used `dynamic_cast`, which we don't prefer. Reason is:
  - I tried to replace ``EventRecorder::_duration_events` (member var, a vector) with  _two_ vectors(one for `SubgDurationEvent` and `OpSeqDurationEvent`) 
  - However this makes the implementation logic of `MDTableWriter`, which heavily use index of one vector. So I decided to keep one vector and `dynamic_cast`. (maybe we can refactor this later)
  - `EventWriter`, which uses `dynamic_cast` does not reduce the _inference_ performance of ONERT.

hyunsik-yoon commented:
@nnfw-bot test cross-runtime-test-debug
@nnfw-bot test cross-runtime-test-release

hyunsik-yoon commented:
@nnfw-bot test cross-runtime-test-debug
@nnfw-bot test cross-runtime-test-release

hyunsik-yoon commented:
@nnfw-bot test cross-runtime-test-debug
@nnfw-bot test cross-runtime-test-release

hyunsik-yoon commented:
@nnfw-bot test cross-runtime-test-release

hyunsik-yoon commented:
symbol for operation have been changed as modified in #5448

wateret commented:
@nnfw-bot test cross-runtime-test-debug

wateret reviewed (approved):
LGTM
